compiler/src/iree/compiler/InputConversion/Common/InputConversionPipeline.cpp

```diff
@@ -0,0 +1,146 @@
+
+#include "iree/compiler/InputConversion/Common/PassDetail.h"
+#include "iree/compiler/InputConversion/Common/Passes.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Tosa/IR/TosaOps.h"
+#include "mlir/IR/BuiltinDialect.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+
+// Dialect specific
+#ifdef IREE_HAVE_MHLO_INPUT
+#include "iree/compiler/InputConversion/MHLO/Passes.h"
+#include "mhlo/IR/hlo_ops.h"
+#endif
+#ifdef IREE_HAVE_TOSA_INPUT
+#include "iree/compiler/InputConversion/TOSA/Passes.h"
+#endif  // IREE_HAVE_TOSA_INPUT
+#ifdef IREE_HAVE_TORCH_INPUT
+#include "iree/compiler/InputConversion/TMTensor/Passes.h"
+#include "torch-mlir-dialects/Dialect/TMTensor/IR/TMTensorDialect.h"
+#endif  // IREE_HAVE_TORCH_INPUT
+
+namespace mlir::iree_compiler {
+
+struct InputConversionPipelinePass
+    : public InputConversionPipelineBase<InputConversionPipelinePass> {
+  void runOnOperation() override;
+};
+
+#if defined(IREE_HAVE_MHLO_INPUT) || defined(IREE_HAVE_TOSA_INPUT)
+
+// All the features seen that should be handled during input conversion.
+struct InputFeatures {
+  // MHLO features.
+  bool hasMHLO = false;
+  bool hasStableHLO = false;
+  // - XLA import features.
+  bool hasTuples = false;
+
+  // TOSA features.
+  bool hasTOSA = false;
+
+  // tm_tensor
+  bool hasTmTensor = false;
+};
+
+static void populateHloFeatures(Operation* op, InputFeatures& features) {
+  features.hasMHLO = true;
+  if (features.hasTuples) return;
+
+  if (auto funcOp = dyn_cast<func::FuncOp>(op)) {
+    FunctionType type = dyn_cast<FunctionType>(funcOp.getFunctionType());
+    for (auto t : type.getResults()) {
+      if (isa<TupleType>(t)) return (void)(features.hasTuples = true);
+    }
+    for (auto t : type.getInputs()) {
+      if (isa<TupleType>(t)) return (void)(features.hasTuples = true);
+    }
+  }
+
+  // Check for tuple operands or results.
+  for (auto t : op->getOperandTypes())
+    if (isa<TupleType>(t)) return (void)(features.hasTuples = true);
+  for (auto t : op->getResultTypes())
+    if (isa<TupleType>(t)) return (void)(features.hasTuples = true);
+}
+
+static void populateFeatures(Operation* op, const Dialect* mhloDialect,
+                             const Dialect* tmTensorDialect,
+                             const Dialect* tosaDialect,
+                             InputFeatures& features) {
+  Dialect* d = op->getDialect();
+  if (d == mhloDialect) return populateHloFeatures(op, features);
+  if (d == tosaDialect) return (void)(features.hasTOSA = true);
+  if (d == tmTensorDialect) return (void)(features.hasTmTensor = true);
+}
+
+void InputConversionPipelinePass::runOnOperation() {
+  ModuleOp module = getOperation();
+  MLIRContext* ctxt = &getContext();
+
+  InputFeatures features;
+
+  const Dialect* mhloDialect = nullptr;
+  const Dialect* tosaDialect = nullptr;
+  const Dialect* tmTensorDialect = nullptr;
+#ifdef IREE_HAVE_MHLO_INPUT
```

Comment:
I am not sure having `ifdef`s in the code is the way we want to support this.....

Reply:
It is how we are currently supporting these. I could switch to string based version (`getLoadedDialect(StringRef name)`) and then just use ifdefs for includes.

Reply:
Yep, the defines let us make each input type (and associated submodule, build, test, etc. deps) optional. `getLoadedDialect` is an interesting idea... that would let us centralize the defines to make the rest of the code a little cleaner.